Add colony event log to Town Hall

ToMe25 opened the pull request:
# Changes proposed in this pull request:
- A colony event log that can be shown in the town hall instead of the permission event log.
- Replace all tabs in java files with spaces.
- Group a few more of the Town Hall lang keys together in the lang file.

Here a screenshot of the colony event log:
![2020-09-10_19 24 56](https://user-images.githubusercontent.com/38815969/92771676-5f259380-f39b-11ea-808a-bf60eb18de8f.png)


Review please

ToMe25 committed:
Initial Partial implementation

 * Add citizen spawn IColonyEvents
 * Add IColonyEvent handling to Town Hall Gui

ToMe25 committed:
Change colony log to new IColonyEventDescription instead of IColonyEvent

 * Add citizen death event
 * Fix some issues
 * undo like half the changes of the last commit
 * The base systems should work now

ToMe25 committed:
Add building events(not fully implemented)

 * replace all tabs with spaces everywhere in the source code.
 * fix death message formatting

ToMe25 committed:
Merge remote-tracking branch 'origin/version/1.15' into
feature/colony_event_log

ToMe25 committed:
Seems like this class got merged in some strange way

This should fix it.

ToMe25 committed:
Add CitizenGrownUpEvent

also make CitizenDiedEvent a subclass of AbstractCitizenEvent(formerly
AbstractCitizenSpawnEvent)

ToMe25 committed:
formatting

ToMe25 committed:
formatting

ToMe25 committed:
Merge branch 'version/1.15' into feature/colony_event_log

ToMe25 committed:
Add a button switching between colony and permission events

Also group some more of the town hall lang entries together, but its
still not nearly all of them.

ToMe25 committed:
Improve building name shown in colony event log

also undo a formatting change in TranslationConstants

ToMe25 committed:
improve translation constants

 * Move the townhall gui translation keys from WindowTownHall to
TranslationConstants
 * add NonNls annotation to raid name TranslationConstants
 * make some formatting more consistent

ToMe25 committed:
remove todo for a better way to detect upgrading

cast work order to WorkOrderBuild instead of WorkOrderBuildBuilding

Raycoms commented:
GUI in screenshot seems to overlap at the bottom

ToMe25 commented:
yes, i forgot to change that.
Also that was a problem even before this pr.

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/colonyEvents/citizenEvents/CitizenBornEvent.java:
```diff
@@ -0,0 +1,59 @@
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.network.PacketBuffer;
+import net.minecraft.util.ResourceLocation;
+
```
Something here seems off

ToMe25 replied:
why? its just three imports and an empty line before the class description.

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/colonyEvents/citizenEvents/CitizenGrownUpEvent.java:
```diff
@@ -0,0 +1,59 @@
+
+import org.jetbrains.annotations.NotNull;
+
+import com.minecolonies.api.util.constant.Constants;
```
also here

ToMe25 replied:
do you mean the empty line?
if yes, thats how eclipse formats imports, i can remove it if you want.

Raycoms replied:
nvm, it was a github problem

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/managers/CitizenManager.java:
```diff
@@ -295,6 +297,11 @@ private ICitizenData spawnCitizenOnPosition(
+        AbstractCitizenEvent spawnEvent = citizenData.isChild() ? new CitizenBornEvent() : new CitizenSpawnedEvent();
+        spawnEvent.setEventPos(spawnPoint);
+        spawnEvent.setCitizenName(citizenData.getName());
+        colony.getEventManager().addEventDescription(spawnEvent);
```
I feel like those things would fit better into the constructor of the event instead of manually setting pos and name.

ToMe25 replied:
thats a relic from a changed implementation, i thought i changed it, but seems like i didn't.

ToMe25 replied:
This time i really changed that, and not just planned to do it xD

Raycoms replied:
:+1: xD

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/managers/EventManager.java:
```diff
@@ -81,6 +87,16 @@ public void addEvent(final IColonyEvent colonyEvent)
     }
 
+    @Override
+    public void addEventDescription(IColonyEventDescription colonyEventDescription) {
```
formatting here is off

ToMe25 replied:
fixed.

ToMe25 committed:
Some fixes

 * fix event log overlapping with gui border
 * move some value initialization to constructors

Raycoms committed:
Merge branch 'version/1.15' into feature/colony_event_log

someaddons commented on src/main/java/com/minecolonies/coremod/colony/managers/CitizenManager.java:
```diff
@@ -295,6 +297,9 @@ private ICitizenData spawnCitizenOnPosition(
         world.addEntity(entity);
 
+        colony.getEventManager().addEventDescription(citizenData.isChild() ? new CitizenBornEvent(spawnPoint, citizenData.getName()) :
+              new CitizenSpawnedEvent(spawnPoint, citizenData.getName()));
```
this is also used for entity respawning, which would trigger the event again

ToMe25 replied:
"entity respawning" as in loading the world again, or recalling, or...?

someaddons replied:
loading the world no, but if its missing or gets recalled to respawn etc, minecraft sometimes looses the entities and we need to respawn them

ToMe25 replied:
if the citizen goes missing, does `citizenData.getEntity()` return null, or an invalid entity?

someaddons replied:
could be either

someaddons replied:
BuildingHome has the child spawn, and the RecruitmentInteraction has the visitors

ToMe25 replied:
so the only solution is to move the event handling to the classes requesting the citizen spawn instead of the spawn handler?

someaddons replied:
well spawning is only the entity, not the citizen itself. Entities are exchangeable/get lost, citizenData is the actualy citizen

ToMe25 replied:
wait, does that mean i can handle that in the CitizenData constructor?

someaddons replied:
no thats called upon loading too :D

ToMe25 replied:
once the crash is fixed i will look more into this.

ToMe25 replied:
seems like the recent citizen loading changes cause world loading to sometimes cause a citizen spawned event, but not always.

ToMe25 replied:
I hope this should be completely done now.
I tested it some time and it seems to fully work.

someaddons commented on src/main/java/com/minecolonies/coremod/colony/managers/EventManager.java:
```diff
@@ -53,6 +59,11 @@
+     * The event descriptions of this colony.
+     */
+    private final LinkedList<IColonyEventDescription> eventDescs = new LinkedList<>();
+
```
ColonyEvents are quite different from event descriptions, those are ongoing timed events for a colony like a pirate raid. I don't think that the event descriptions belong into here at all, better move them to their own file.

ToMe25 replied:
I have them here, because that way i can reuse code from the EventManager, but i can move them to their own class.

someaddons replied:
ye make the own manager for it, like ColonyHistoricEventManager or os

Raycoms replied:
Maybe call it ColonyLogManager?

Raycoms replied:
it's kinda a log

ToMe25 replied:
I called it EventDescriptionManager, but i want to test it before pushing it, and i can't do that because of the crash.

ToMe25 replied:
this should be done now.

someaddons reviewed (changes requested):
I like the idea of having an eventlog for the colony, but would like to see some changes to this still

ToMe25 committed:
Merge branch 'version/1.15' into feature/colony_event_log

ToMe25 committed:
Move event description handling to its own class

marchermans commented on src/api/java/com/minecolonies/api/colony/colonyEvents/descriptions/IBuildingEventDescription.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Event description for building events.
+ */
+public interface IBuildingEventDescription extends IColonyEventDescription
```
Please rename the methods in this interface to be more accurate.
Aka `getBuildingName` instead of `getBuilding`

ToMe25 replied:
done.

marchermans commented on src/api/java/com/minecolonies/api/colony/colonyEvents/descriptions/IColonyEventDescription.java:
```diff
@@ -0,0 +1,79 @@
+/**
+ * Description for an event that happened in the colony.
+ */
+public interface IColonyEventDescription
```
`extends INBTSerializable<CompoundNBT>` should be used here instead of the custom methods.

ToMe25 replied:
done.

ToMe25 replied:
This is now also done for `IColonyEvent`.

marchermans commented on src/api/java/com/minecolonies/api/colony/colonyEvents/descriptions/NBTTags.java:
```diff
@@ -0,0 +1,13 @@
+/**
+ * NBT tags used in colony events.
+ */
+public abstract class NBTTags
```
Use the `NBTTagConstantsClass`

ToMe25 replied:
ok, i did it like this because that is how `IColonyEvent` handles its tags.

marchermans replied:
Then fix that too :smile:

ToMe25 replied:
done.

marchermans commented on src/api/java/com/minecolonies/api/colony/colonyEvents/registry/ColonyEventDescriptionTypeRegistryEntry.java:
```diff
@@ -0,0 +1,65 @@
+/**
+ * This is the colonies event registry entry class, used for registering any colony related events. Takes a function of colony, nbt to create the right event object.
+ */
+public class ColonyEventDescriptionTypeRegistryEntry extends ForgeRegistryEntry<ColonyEventDescriptionTypeRegistryEntry>
```
Rework this class so that it does not expose the inner function callbacks.

ToMe25 replied:
This is how its handled for `IColonyEvent`s tho, does it really have to change?

marchermans replied:
Yes.
The functions are inner information that is not supposed to be exposed to the external world.

Please fix this in IColonyEvent as well.

ToMe25 replied:
done.

marchermans commented on src/api/java/com/minecolonies/api/colony/managers/interfaces/IEventDescriptionManager.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Interface for the event description manager, the event description manager deals the colony event log events.
+ */
+public interface IEventDescriptionManager
```
Again missing the `INBTSerializable<CompoundNBT>` instead of using the custom methods.

ToMe25 replied:
done.

marchermans commented on src/main/java/com/minecolonies/apiimp/CommonMinecoloniesAPIImpl.java:
```diff
@@ -38,20 +39,21 @@
-    private final  IInteractionResponseHandlerDataManager          interactionDataManager = new InteractionResponseHandlerManager();
-    private        IForgeRegistry<ColonyEventTypeRegistryEntry>    colonyEventRegistry;
-    private static IGlobalResearchTree                             globalResearchTree     = new GlobalResearchTree();
+    private final  IColonyManager                                          colonyManager          = new ColonyManager();
+    private final  ICitizenDataManager                                     citizenDataManager     = new CitizenDataManager();
+    private final  IMobAIRegistry                                          mobAIRegistry          = new MobAIRegistry();
+    private final  IPathNavigateRegistry                                   pathNavigateRegistry   = new PathNavigateRegistry();
+    private        IForgeRegistry<BuildingEntry>                           buildingRegistry;
+    private final  IBuildingDataManager                                    buildingDataManager    = new BuildingDataManager();
+    private final  IJobDataManager                                         jobDataManager         = new JobDataManager();
+    private final  IGuardTypeDataManager                                   guardTypeDataManager   = new com.minecolonies.coremod.colony.buildings.registry.GuardTypeDataManager();
+    private        IForgeRegistry<JobEntry>                                jobRegistry;
+    private        IForgeRegistry<GuardType>                               guardTypeRegistry;
+    private        IForgeRegistry<InteractionResponseHandlerEntry>         interactionHandlerRegistry;
+    private final  IInteractionResponseHandlerDataManager                  interactionDataManager = new InteractionResponseHandlerManager();
+    private        IForgeRegistry<ColonyEventTypeRegistryEntry>            colonyEventRegistry;
+    private        IForgeRegistry<ColonyEventDescriptionTypeRegistryEntry> colonyEventDescriptionRegistry;
+    private static IGlobalResearchTree                                     globalResearchTree     = new GlobalResearchTree();
```
What happened here?

ToMe25 replied:
formatting changes happened here, there were spaces to make all the names start at the same point, and since the new one has a longer type i moved all the var names.

marchermans replied:
Okey

marchermans commented on src/main/java/com/minecolonies/coremod/colony/managers/CitizenManager.java:
```diff
@@ -282,6 +284,9 @@ private ICitizenData spawnCitizenOnPosition(
         entity.setPosition(spawnPoint.getX() + HALF_BLOCK, spawnPoint.getY() + SLIGHTLY_UP, spawnPoint.getZ() + HALF_BLOCK);
         world.addEntity(entity);
 
+        colony.getEventDescriptionManager().addEventDescription(citizenData.isChild() ? new CitizenBornEvent(spawnPoint, citizenData.getName()) :
```
Yeah you will need to modify this.
This call location is less then appropriate.

Find the code that actually triggers the birth and fire it from there.

ToMe25 replied:
Addons already told me this, i will fix this once i continue working on this pr later today.

ToMe25 replied:
done.

marchermans reviewed (changes requested):
See comments on the relevant files.

ToMe25 committed:
Merge branch 'version/1.15' into feature/colony_event_log

ToMe25 committed:
add a missing param annotation

ToMe25 committed:
Fix citizen event descriptions getting created on world loading

 * Also merege NBTTags class into NbtTagConstants
 * make visitor spawns cause new VisitorSpawnedEvents instead of
CitizenSpawnedEvents
 * make IColonyEventDescription and IEventDescriptionManager extend
INBTSerializable
 * don't expose deserialization function in ColonyEventTypeRegistryEntry
and ColonyEventDescriptionTypeRegistryEntry

ToMe25 committed:
Merge branch 'version/1.15' into feature/colony_event_log

ToMe25 commented:
This is not entirely done, i missed a small thing :)

ToMe25 committed:
Fix EventDescriptionManager serialization

 * Also fix a comment in VisitorSpawnedEvent
 * also make IColonyEvent implement INBTServializable

ToMe25 commented:
I think this pr is done now, somehow my citizens are invisible now, but because it happened after merging version/1.15 and not after anything i did i doubt thats caused by anything in this pr.

ToMe25 committed:
Fix some formatting

ToMe25 committed:
Merge branch 'version/1.15' into feature/colony_event_log
